module/validation/receipt_validator.go

```diff
@@ -160,64 +161,203 @@ func (v *receiptValidator) Validate(receipts []*flow.ExecutionReceipt) error {
 	// reference each other.
 	// ATTENTION: Here we assume that ER is valid, this lookup can return a result which is actually invalid.
 	// Eventually invalid result will be detected and fail the whole validation.
-	previousResult := func(executionResult *flow.ExecutionResult) (*flow.ExecutionResult, error) {
-		prevResult, found := payloadExecutionResults[executionResult.PreviousResultID]
+	fetchResult := func(previousResultID flow.Identifier) (*flow.ExecutionResult, error) {
+		prevResult, found := payloadExecutionResults[previousResultID]
 		if found {
 			return prevResult, nil
 		}
 
-		return v.previousResult(executionResult)
+		return v.fetchResult(previousResultID)
 	}
 
 	for i, r := range receipts {
-		err := v.validate(r, previousResult)
+		prevResult, err := fetchResult(r.ExecutionResult.PreviousResultID)
+		if err != nil {
+			return err
+		}
+
+		err = v.validateReceipt(r.Meta(), r.ExecutionResult.BlockID)
 		if err != nil {
 			// It's very important that we fail the whole validation if one of the receipts is invalid.
 			// It allows us to make assumptions as stated in previous comment.
 			return fmt.Errorf("could not validate receipt %v at index %d: %w", r.ID(), i, err)
 		}
+		err = v.validateResult(&r.ExecutionResult, prevResult)
```

Comment:
How about this:

```
for each receipt:
  if the result can been found in the storage:
    continue // because results in storage must have been validated
  if the result is a duplication in the payload:
    continue // because a duplicated result must have been validated.
  add the result to a unique result map
  find its previous result
  validate the result with the previous result // validate the result first, before validating the receipt
  validate the receipt meta
```

Reply:
Check [this](https://github.com/onflow/flow-go/pull/462/files#r584048825) comment to see my reasoning.

Reply:
The reason I suggest this flow is that I think we'd better validate the result first before validating the receipt, because if the result is bad, we wouldn't waste effort validating the signature from the receipt, as validating the signature is expensive.